src/vapid-helper.js

```diff
@@ -71,18 +71,22 @@ function validateSubject(subject) {
   }
 
   if (typeof subject !== 'string' || subject.length === 0) {
-    throw new Error('The subject value must be a string containing a URL or '
+    throw new Error('The subject value must be a string containing an https: URL or '
     + 'mailto: address. ' + subject);
   }
 
-  if (subject.indexOf('mailto:') !== 0) {
-    const subjectParseResult = url.parse(subject);
-    if (!subjectParseResult.hostname) {
-      throw new Error('Vapid subject is not a url or mailto url. ' + subject);
-    } else if (subjectParseResult.hostname === 'localhost' && subjectParseResult.protocol === 'https:') {
-      console.warn('VAPID subject points to a localhost web URI, which is unsupported by Apple\'s push notification '
-      + 'server and will result in a BadJwtToken error when sending notifications.');
+  try {
+    const subjectParseResult = new URL(subject);
+    if (!['https:', 'mailto:'].includes(subjectParseResult.protocol)) {
+      throw new Error('Vapid subject is not an https: or mailto: URL. ' + subject);
     }
+    if (subjectParseResult.hostname === 'localhost') {
```

Comment:
Doesn't this also warn in case of mailto?

Reply:
No; the WHATWG URL parser doesn't put the mailto: hostname into the hostname property, it ends up in the `pathname` property (by design). See [here](https://runkit.com/delfuego/url-parsing-next-gen).